test/sanity/code-smell/rstcheck.py

```diff
@@ -4,59 +4,170 @@
 import re
 import subprocess
 import sys
+import typing as t
+from shlex import join as _shlex_join
```

Comment:
This seems like a lot of unnecessary refactoring and excessive changes for what is actually needed. A much simpler change applies the directive ignore to the relevant files:

```diff
diff --git a/test/sanity/code-smell/rstcheck.py b/test/sanity/code-smell/rstcheck.py
index 4b5a43e69c..6efd2ff85c 100644
--- a/test/sanity/code-smell/rstcheck.py
+++ b/test/sanity/code-smell/rstcheck.py
@@ -20,21 +20,33 @@ def main():
         '-c', 'import rstcheck; rstcheck.main();',
         '--report', 'warning',
         '--ignore-substitutions', ','.join(ignore_substitutions),
-    ] + paths
+    ]
 
-    process = subprocess.run(cmd,
-                             stdin=subprocess.DEVNULL,
-                             stdout=subprocess.PIPE,
-                             stderr=subprocess.PIPE,
-                             check=False,
-                             )
+    docsite_path = 'docs/docsite/rst/cli/'
 
-    if process.stdout:
-        raise Exception(process.stdout)
+    commands = [
+        ['--ignore-directives', 'ansible-cmd-docs'] + [path for path in paths if path.startswith(docsite_path)],
+        [path for path in paths if not path.startswith(docsite_path)]
+    ]
+
+    errors = ''
+
+    for command in commands:
+        process = subprocess.run(cmd + command,
+                                 stdin=subprocess.DEVNULL,
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.PIPE,
+                                 check=False,
+                                 )
+
+        if process.stdout:
+            raise Exception(process.stdout)
+
+        errors += process.stderr.decode(encoding)
 
     pattern = re.compile(r'^(?P<path>[^:]*):(?P<line>[0-9]+): \((?P<level>INFO|WARNING|ERROR|SEVERE)/[0-4]\) (?P<message>.*)$')
 
-    results = parse_to_list_of_dict(pattern, process.stderr.decode(encoding))
+    results = parse_to_list_of_dict(pattern, errors)
 
     for result in results:
         print('%s:%s:%s: %s' % (result['path'], result['line'], 0, result['message']))
```

Reply:
@mattclay this approach makes it harder to follow and doesn't process some of the failures as gracefully. So I'd like to keep it refactored. Would it be okay if I separated out refactoring into a dedicated PR?

Reply:
@webknjaz Can you share example(s) of the kinds of failures you're referring to?

Reply:
Here's my concerns about the original implementation and the patch proposal:
1) An unspecified generic exception being used all over the place makes it hard to reason about the control flow
2) An unobvious structuring side-effect: unexpected output of the first subprocess invocation prevents us from seeing if the following invocation has anything useful to report
3) Unnecessary Python 2 compatibility hacks (working with bytes, old-style string interpolation etc)
4) It is poorly structured (things like a regexp being declared in an arbitrary scope that is not related to using it directly) which manifests as parts being tightly coupled for no good reason — it's just not clean code

Reply:
> 1. An unspecified generic exception being used all over the place makes it hard to reason about the control flow

There are only two exceptions raised in the current implementation, both of which terminate the script with an actionable error message. Replacing that with six raised exceptions and three exception handlers makes the code significantly more complex.

> 2. An unobvious structuring side-effect: unexpected output of the first subprocess invocation prevents us from seeing if the following invocation has anything useful to report

I don't consider this an issue worth increasing code complexity over. I can't recall the last time I encountered unexpected output from the subprocess in this test. If it does occur, there's nothing wrong with having to resolve that issue before seeing further output.

> 3. Unnecessary Python 2 compatibility hacks (working with bytes, old-style string interpolation etc)

There's a single use of `bytes.decode` which could be replaced by passing `text=True` to `subprocess.run`. There are two strings that could be updated to use f-strings. There's nothing broken about either of those, and modernizing them only involves removing two lines of code, adding one, and changing one.

> 4. It is poorly structured (things like a regexp being declared in an arbitrary scope that is not related to using it directly) which manifests as parts being tightly coupled for no good reason — it's just not clean code

As a fairly trivial program, there's not much structure that could be reasonably simplified. For the regex example given, that could be handled by moving the regex into the function. Again, it's not broken, and it has a debatable impact on readability.

If there was good reason to refactor the test (I don't think there is), then those changes would be more appropriate in separate PR.

Overall, as a currently very simple sanity test, I think there's much more value in a small, simple implementation  than one nearly three times the size that adds significant complexity. I realize it's a matter of perspective, but I can much more easily understand the current implementation (and my patched version) than the more complex one in this PR.

Reply:
I understand that some of these are perspective-dependent and wouldn't be changing things if I didn't believe this is the right thing to do. Obviously, you can more easily understand your version and I can more easily understand my version. This is also because I've spent some time thinking things through just recently and have a fresh memory of the first impression and things I wanted to improve.

Another bit I noticed missing in your patch (that I recall trying to solve in my refactoring cleanly) is that there's already some structure to extra arguments for `--ignore-substitutions` but not for the new thing. This brings in some new levels of messiness.

Now that I've had a chance to think about this again, I looked into my implementation and remembered another thing that I solved — calling the check against a single document. I know that with your implementation, it'll always fail because you'll be passing an empty list of targets into one of the subcommands which makes it error out.

FWIW as you've started suggesting ways to address my concerns, I got a feeling that you're starting to go through a similar thought process and given enough iterations would end up with something close to what I did. On the high level, at least..

I've already suggested separating the refactoring out into a separate PR specifically because I think there's a good reason for doing this.

Reply:
Lets go with a minimal set of test changes for this PR. Any refactoring can be considered in a separate PR later.

I updated my patch (against devel) to account for the one file scenario you mentioned.

```diff
diff --git a/test/sanity/code-smell/rstcheck.py b/test/sanity/code-smell/rstcheck.py
index 4b5a43e69c..13b619c081 100644
--- a/test/sanity/code-smell/rstcheck.py
+++ b/test/sanity/code-smell/rstcheck.py
@@ -20,21 +20,35 @@ def main():
         '-c', 'import rstcheck; rstcheck.main();',
         '--report', 'warning',
         '--ignore-substitutions', ','.join(ignore_substitutions),
-    ] + paths
+    ]
 
-    process = subprocess.run(cmd,
-                             stdin=subprocess.DEVNULL,
-                             stdout=subprocess.PIPE,
-                             stderr=subprocess.PIPE,
-                             check=False,
-                             )
+    docsite_path = 'docs/docsite/rst/cli/'
 
-    if process.stdout:
-        raise Exception(process.stdout)
+    scenarios = [
+        (['--ignore-directives', 'ansible-cmd-docs'], [path for path in paths if path.startswith(docsite_path)]),
+        ([], [path for path in paths if not path.startswith(docsite_path)]),
+    ]
+
+    commands = [cmd + options + files for options, files in scenarios if files]
+
+    errors = ''
+
+    for command in commands:
+        process = subprocess.run(command,
+                                 stdin=subprocess.DEVNULL,
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.PIPE,
+                                 check=False,
+                                 )
+
+        if process.stdout:
+            raise Exception(process.stdout)
+
+        errors += process.stderr.decode(encoding)
 
     pattern = re.compile(r'^(?P<path>[^:]*):(?P<line>[0-9]+): \((?P<level>INFO|WARNING|ERROR|SEVERE)/[0-4]\) (?P<message>.*)$')
 
-    results = parse_to_list_of_dict(pattern, process.stderr.decode(encoding))
+    results = parse_to_list_of_dict(pattern, errors)
 
     for result in results:
         print('%s:%s:%s: %s' % (result['path'], result['line'], 0, result['message']))
```